Add support for WCH CH32V series I2C controller

andrei-edward-popa opened the pull request:
This PR adds a new I2C driver for the WCH CH32V series microcontrollers.
The driver implements standard I2C master functionality using the
hardware peripheral available on CH32V003 and similar WCH MCUs.

Features:
- Supports standard (100 kHz) and fast mode (400 kHz) speeds.
- Implements interrupt-driven transfers for efficient communication.

Tested on:
- ch32v003evt
- ch32v003f4p6_dev_board

andrei-edward-popa committed:
drivers: i2c: added wch i2c driver

added i2c driver for wch platforms

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
dts: riscv: wch: added i2c node

added i2c node for ch32v003

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
boards: wch: ch32v003f4p6_dev_board: enable i2c and define i2c pins

enabled i2c for ch32v003f4p6_dev_board
added default i2c pins for ch32v003f4p6_dev_board
added i2c as supported

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
dts: bindings: i2c: added bindings for wch

added i2c bindings for wch platforms

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

VynDragon commented:
Thank you! I will take a look at this later and check it works with the other SoCs.

andrei-edward-popa commented:
> So how did you actually test this? CH32V203 didnt work, i tried CH32V003 and found out neither bme280 or display (ssd1315) sample fit...

Hi @VynDragon! I created some samples that fits into ch32v003 like i2c scan and drivers for lcd 1602, ds3231 and mp6050 here: https://github.com/andrei-edward-popa/zephyr/tree/wch_examples/samples/boards/wch. Can you try one of those?

miggazElquez commented:
I tested it on the ch32v303 and it works fine. I didn't test the interrupts.

andrei-edward-popa commented:
> I tested it on the ch32v303 and it works fine. I didn't test the interrupts.

Hi @miggazElquez! Thank you for checking it on ch32v303, if you do more tests and found some issues let me know.

VynDragon commented:
I must be doing something wrong, i see no activity at all on either 203 or 003.

andrei-edward-popa commented:
> I must be doing something wrong, i see no activity at all on the bus on either 203 or 003.

I guess you are using ch32v003evt, I didn't create the pinctrl.dtsi and enabled the i2c in dts for it, did you do it? If yes, did you enable gpioc?

VynDragon commented:
Yea, Scratch the ch32v003 part, just got it working.

VynDragon commented:
Okay so still no activity with the display sample, but the scan sample shows activity on the bus (that looks like toggling pins and off though), but doesn't find anything.

andrei-edward-popa commented:
> Okay so still no activity with the display sample, but the scan sample shows activity on the bus (that looks like toggling pins and off though), but doesn't find anything.

Can you tell me what board are you using? I want to order one also.

VynDragon commented:
my dts overlay for the example code is: 
```
#include <dt-bindings/pinctrl/ch32v20x_30x-pinctrl.h>
#include <zephyr/dt-bindings/i2c/i2c.h>

&pinctrl {
	i2c1_default: i2c1_default {
		group1 {
			pinmux = <I2C1_SDA_PB7_0>, <I2C1_SCL_PB6_0>;
			bias-pull-up;
			drive-open-drain;
			output-high;
			slew-rate = "max-speed-10mhz";
		};
	};
};


/ {
	aliases {
		i2c-0 = &i2c1;
	};
};

&i2c1 {
	status = "okay";

	pinctrl-0 = <&i2c1_default>;
	pinctrl-names = "default";
	clock-frequency = <I2C_BITRATE_STANDARD>;
};

```

i have added  this:

```
		i2c1: i2c@40005400 {
			compatible = "wch,i2c";
			#address-cells = <1>;
			#size-cells = <0>;
			reg = <0x40005400 0x400>;
			clocks = <&rcc CH32V20X_V30X_CLOCK_I2C1>;
			interrupt-parent = <&pfic>;
			interrupts = <33>,<34>;
		};
```

to soc dts,
GPIOB (i2c1 on 20x) is already enabled, and i have checked the GPIO config is supposed to be correct.

VynDragon commented:
> Can you tell me what board are you using? I want to order one also.

Weactstudio ch32v203 blue pill plus (https://github.com/zephyrproject-rtos/zephyr/pull/87490)

VynDragon commented:
> I tested it on the ch32v303 and it works fine. I didn't test the interrupts.

Did you actually test functionality or just build? 203 and 303 should be similar enough that if it works for you it should work for me.

Anyway bus looks like that:
 
![image](https://github.com/user-attachments/assets/af52f337-5b34-4568-ac33-3fbba6ad73a4)
![image](https://github.com/user-attachments/assets/81eb5553-fb07-471b-8ad8-cd4aa5af6d9e) (all identical like this)

andrei-edward-popa commented:
> Figured my error, I read the wrong column for Interrupt numberm it works! (33 34 should be 47 48)
> 
> Edit: Display is still a no-go though, I have to ask drivers through i2c should work.

That's what I wanted to say, I looked into the RM and I saw 31 and 32 for I2C1 and 33 and 34 for I2C2. Where did you find 47 and 48? I think they should be 31 and 32.

VynDragon commented:
> > Figured my error, I read the wrong column for Interrupt numberm it works! (33 34 should be 47 48)
> > Edit: Display is still a no-go though, I have to ask drivers through i2c should work.
> 
> That's what I wanted to say, I looked into the RM (CH32FV2x_V3xRM.PDF) and I saw 31 and 32 for I2C1 and 33 and 34 for I2C2. Where did you find 47 and 48? I think they should be 31 and 32.

You are also in the wrong column like i was :)
![image](https://github.com/user-attachments/assets/5e38fc19-6498-439b-bc98-1d2642b554a6)
![image](https://github.com/user-attachments/assets/c0862f9a-ac62-46ec-b527-b2296c8257b3)


anyway here is the bus action: 
![image](https://github.com/user-attachments/assets/f77bf145-5e28-4c14-b02f-0b877af27582)
Seems it sends the first command, but returns not success to the display driver, possibly due to lack of burst support:
```i2c_burst_write_dt(&config->bus.i2c,
				  command ? SSD1306_CONTROL_ALL_BYTES_CMD :
				  SSD1306_CONTROL_ALL_BYTES_DATA,
				  buf, len);```
				  data send by this command is 0x0 then 0xAE

andrei-edward-popa commented:
> > > Figured my error, I read the wrong column for Interrupt numberm it works! (33 34 should be 47 48)
> > > Edit: Display is still a no-go though, I have to ask drivers through i2c should work.
> > 
> > 
> > That's what I wanted to say, I looked into the RM (CH32FV2x_V3xRM.PDF) and I saw 31 and 32 for I2C1 and 33 and 34 for I2C2. Where did you find 47 and 48? I think they should be 31 and 32.
> 
> You are also in the wrong column (or table? maybe you are lookingat the ARM version ones) like i was :) ![image](https://private-user-images.githubusercontent.com/3176238/429137020-5e38fc19-6498-439b-bc98-1d2642b554a6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1MjY5MTAsIm5iZiI6MTc0MzUyNjYxMCwicGF0aCI6Ii8zMTc2MjM4LzQyOTEzNzAyMC01ZTM4ZmMxOS02NDk4LTQzOWItYmM5OC0xZDI2NDJiNTU0YTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDFUMTY1NjUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWJmYmQ0NWEyNjA2Zjc0MjM4ODVlZWE3NzQyOTAwZTNiNTRmNzI4YTQwYzNiZGQyMjY2NmNjOTJmODRmZmE4NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ywyWv5B-QxIUyKk1fSk0kEElD0ZW_Ny6fHKYKXIwS8U) ![image](https://private-user-images.githubusercontent.com/3176238/429137137-c0862f9a-ac62-46ec-b527-b2296c8257b3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1MjY5MTAsIm5iZiI6MTc0MzUyNjYxMCwicGF0aCI6Ii8zMTc2MjM4LzQyOTEzNzEzNy1jMDg2MmY5YS1hYzYyLTQ2ZWMtYjUyNy1iMjI5NmM4MjU3YjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDFUMTY1NjUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmYxNjNkNTAxMjhlZjQ0MTRiMzg0ODIwZDZmZmUyMjFlOWRlYTQ4MmIxZTI5NGNmZGE3ZTRkNDY3NGMzNjlmNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.xD-JLtMr9rmFM8DuqdLH0tLODhRrdhJs4CNonWZsdww)
> 
> anyway here is the bus action: ![image](https://private-user-images.githubusercontent.com/3176238/429136880-f77bf145-5e28-4c14-b02f-0b877af27582.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1MjY5MTAsIm5iZiI6MTc0MzUyNjYxMCwicGF0aCI6Ii8zMTc2MjM4LzQyOTEzNjg4MC1mNzdiZjE0NS01ZTI4LTRjMTQtYjAyZi0wYjg3N2FmMjc1ODIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDFUMTY1NjUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWZmMzFjNDUxYTA5ZDliNWEzNzBmMTY2ZjFjNzk3MGExMWIyZmRhMGE4ZGYzNjBkYjY4YTZkMGZiNDIzZDdmOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.htj1Ytyny8tHFoaX0Oz5F-QHZZqHkYi0Du6K5N4uAFQ) Seems it sends the first command, but returns not success to the display driver:
> 
> ```
> i2c_burst_write_dt(&config->bus.i2c,
> 				  command ? SSD1306_CONTROL_ALL_BYTES_CMD :
> 				  SSD1306_CONTROL_ALL_BYTES_DATA,
> 				  buf, len);
> ```
> 
> data send by this command is 0x0 then 0xAE so otherwise it is correct.

Yeah, I was in the table for F version, not V :))

VynDragon commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,431 @@
+		}
+	}
+
+	(void)regs->STAR1;
```
Nit: probably unecessary, dont know if other reviewers would accept it

andrei-edward-popa replied:
Why unnecessary? The manual says to read STAR1/STAR2 in some situations.

VynDragon replied:
I mean the (void)regs->star1, you dont need to mark it as unused as it's not a function argument

VynDragon commented on dts/bindings/i2c/wch,i2c.yaml:
```diff
@@ -0,0 +1,16 @@
+properties:
+  reg:
+    required: true
+  clocks:
```
nordicjm would say you need newlines between the properties.

andrei-edward-popa replied:
I put newlines

VynDragon reviewed (changes requested):
A few changes you can already while solving the current issue (s?)

Adding a i2c binding to ch32v003evt would be nice too (but wont fail review for that)

VynDragon commented:
Also the failure that returns error is the semaphore timeing out in begin transfer, hope this is helpful enough to dtermine what's wrong until you get a board that can run some of the driver samples/things.

VynDragon commented:
> > Also the failure that returns error is the semaphore timeing out in begin transfer, hope this is helpful enough to dtermine what's wrong until you get a board that can run some of the driver samples/things.
> 
> Thank you for your comments and findings, @VynDragon! I will try to find the issue tomorrow or in some hours and resolve all the comments. So, to be clear, after you send the write command, only 0xAE must be send, but a 0x00 is send before it?

No, that's right (on solomon display controllers that have i2c, 0x0 is to say 'this is commands following', as opposed to 0x40 which means 'this is data following', and then 0xAE is the display blanking command, first sent).
What is not right is the data is transfered as expected but the i2c driver reports that it failed (due to the aformentionned semaphore timeout).

miggazElquez commented:
> Did you actually test functionality or just build?

I did test it, but with a very simple device, [the nunchuk](https://docs.zephyrproject.org/latest/build/dts/api/bindings/input/nintendo%2Cnunchuk.html), and the sample `input_dump`. It's a simple device, the code just does one write and one read on a loop.

VynDragon commented:
> > Did you actually test functionality or just build?
> 
> I did test it, but with a very simple device, [the nunchuk](https://docs.zephyrproject.org/latest/build/dts/api/bindings/input/nintendo%2Cnunchuk.html), and the sample `input_dump`. It's a simple device, the code just does one write and one read on a loop.

I see, maybe it is a indication of what is failing and what is not, would you mind trying a sample like sensor/bme280 or drivers/display with a i2c display of your choice to see if you can reproduce the issue on ch32v303?

andrei-edward-popa commented:
@VynDragon I found something strange with this I2C, which is reproducable on ch32v003 also and it seems not to be a software issue. Basically, when board is powered on, SDA or SCL or both sometimes stays low, which is weird. When both are high, the communication works as expected, but, sometimes, after some power cycles or reflashing with the same elf, it turns out that SDA/SCL are low instead of high and the timeout occurs. Can you try to put this in your application to see if both lines are high and , if they are, if the display works?

const struct device *gpioc = DEVICE_DT_GET(DT_NODELABEL(gpioc));
`const struct device *gpioc = DEVICE_DT_GET(DT_NODELABEL(gpioc));`
`printf("%d\n", gpio_pin_get(gpioc, 1));`
`printf("%d\n", gpio_pin_get(gpioc, 2));`

VynDragon commented:
> @VynDragon I found something strange with this I2C, which is reproducable on ch32v003 also and it seems not to be a software issue. Basically, when board is powered on, SDA or SCL or both sometimes stays low, which is weird. When both are high, the communication works as expected, but, sometimes, after some power cycles or reflashing with the same elf, it turns out that SDA/SCL are low instead of high and the timeout occurs. Can you try to put this in your application to see if both lines are high and , if they are, if the display works?
> 
> `const struct device *gpioc = DEVICE_DT_GET(DT_NODELABEL(gpioc));` `printf("%d\n", gpio_pin_get(gpioc, 1));` `printf("%d\n", gpio_pin_get(gpioc, 2));`

Trace says i2c pins are pulled up correctly before transfer, but not after transfer 
![image](https://github.com/user-attachments/assets/937e2e4d-3006-4f0f-a3df-76624382c85f)

andrei-edward-popa committed:
boards: wch: ch32v003evt: enable i2c and define i2c pins

enabled i2c for ch32v003evt
added default i2c pins for ch32v003evt
added i2c as supported

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa commented:
> > @VynDragon I found something strange with this I2C, which is reproducable on ch32v003 also and it seems not to be a software issue. Basically, when board is powered on, SDA or SCL or both sometimes stays low, which is weird. When both are high, the communication works as expected, but, sometimes, after some power cycles or reflashing with the same elf, it turns out that SDA/SCL are low instead of high and the timeout occurs. Can you try to put this in your application to see if both lines are high and , if they are, if the display works?
> > `const struct device *gpioc = DEVICE_DT_GET(DT_NODELABEL(gpioc));` `printf("%d\n", gpio_pin_get(gpioc, 1));` `printf("%d\n", gpio_pin_get(gpioc, 2));`
> 
> Trace says i2c pins are pulled up correctly before transfer, but not after transfer (neverending stop condition?) ![image](https://private-user-images.githubusercontent.com/3176238/429600194-937e2e4d-3006-4f0f-a3df-76624382c85f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM2Mjg5NTEsIm5iZiI6MTc0MzYyODY1MSwicGF0aCI6Ii8zMTc2MjM4LzQyOTYwMDE5NC05MzdlMmU0ZC0zMDA2LTRmMGYtYTNkZi03NjYyNDM4MmM4NWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDJUMjExNzMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjI0NDBkYWI1OGI2MzgzYWIzZGYyOGRiOGMxOTdhMmI2MjY5ZTRmMzA0ZGM2NmViOWM1NGRjMmE2YTQ3NTFhNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.4jV3Zihp0DIFnIA2bJXaVtuOjz-nan0w3Gdj6Uvp2r4)

@VynDragon I have added a method to reset i2c at init and also changed something that maybe can solve the problem that you have. Please, try it when you can. Thank you!

VynDragon commented:
> @VynDragon I have added a method to reset i2c at init and also changed something that maybe solves the problem that you have. Please, try it when you can. Thank you!

The result at end is same that before, but now it gives a wrong start condition earlier and doesnt transmit the right data. Just wait until you get ch32v203, starting from the initial version, this version has issues (providing scl and sda like it is now to the driver is a problem, if there is a issue with gpio/pinctrl do report and ill try to fix it in gpio and pinctrl) while initial version was cleaner.

Alternatively, you may submit it as a ch32v003 only driver, but if semaphore isnt cleared correctly on ch32v203, it may not be on ch32v003 as well.

VynDragon commented:
Ok then, great!

If i may give a couple recommendations, for digital protocol like this a cheap logic analyzer is sufficient, for example one like this: https://aliexpress.com/item/1005003649856071.html, this one is good up to USB full speed speeds (i have used mine to write a udc driver). For debugging i2c I typically use displays (because i have a lot of those and they enable confirming tx without needing rx, and some are switchable between i2c and spi so enable debugging i2c, spi, and mipi dbi) and BME280 (which is convenient because it does tx, rx, i2c, and spi, you do need one with the select pin broken out though).

andrei-edward-popa commented:
> Ok then, great!
> 
> If i may give a couple recommendations, for digital protocol like this a cheap logic analyzer is sufficient, for example one like this: https://aliexpress.com/item/1005003649856071.html, this one is good up to USB full speed speeds (i have used mine to write a udc driver). For debugging i2c I typically use displays (because i have a lot of those and they enable confirming tx without needing rx, and some are switchable between i2c and spi so enable debugging i2c, spi, and mipi dbi) and BME280 (which is convenient because it does tx, rx, i2c, and spi, you do need one with the select pin broken out though).

I have a digilent digital discovery also in that place, I will try it, thank you @VynDragon :)

nzmichaelh commented:
Neat! FYI, I started a driver at https://github.com/nzmichaelh/zephyr/commit/1c5517654e23dd8faa07da0367badbf899ebbebf but never sent it upstream. Feel free to take anything you want from it.

nzmichaelh commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+
+	if (status & I2C_STAR1_SB) {
+		wch_i2c_handle_start_bit(dev);
+	} else if (status & I2C_STAR1_ADDR) {
```
Are these status bits exclusive? i.e. can you use `else if`?

andrei-edward-popa replied:
If SB for example comes, we need to wait for ADDR interrupt, even if the ADDR bits are set AFAIU, so `else if` is good here.

nzmichaelh commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+	wch_i2c_config_interrupts(regs, true);
+
+	k_sched_lock();
+	while (atomic_get(&data->xfer_done) == 0) {
```
Can this yield in some way? It seems a pity to block the CPU while waiting ~1 ms for an I2C transfer to complete.

andrei-edward-popa replied:
Thank you so much for the review, I will do all the changes on sunday/monday. Regarding this, it was still a WIP and I will change it.

andrei-edward-popa replied:
Outdated: I'm using a semaphone now.

nzmichaelh commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+	uint16_t freq_range;
+
+	if (bitrate <= I2C_BITRATE_STANDARD) {
+		clock_config = MAX((uint16_t)(clock_rate / (bitrate << 1)), 4);
```
nit: for consistency with the (bitrate * 3) below, use (bitrate * 2) here.

andrei-edward-popa replied:
Done

nzmichaelh commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+	}
+
+	if (!(I2C_MODE_CONTROLLER & dev_config)) {
+		return -EINVAL;
```
Note that drivers use a mix of EINVAL and ENOTSUP. I'd use ENOTSUP here as the configuration is valid for the hardware, but it hasn't been implemented.

andrei-edward-popa replied:
Done

nzmichaelh commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+	regs->CTLR1 &= ~I2C_CTLR1_PE;
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
```
wch_i2c_configure_timing supports more modes than listed here. Call wch_i2c_configure_timing directly and let it return an error on unsupported.

I'd also drop the I2C_SPEED_STANDARD -> 100'000 conversion, as wch_i2c_configure_timing then interprets the parameter as a I2C_SPEED_STANDARD etc, so there's more of a chance of introducing bugs through a bad conversion.

andrei-edward-popa replied:
Done

nzmichaelh commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+	if (err != 0) {
+		return err;
+	}
+
```
Also check and reject I2C_ADDR_10_BITS

andrei-edward-popa replied:
Done

nzmichaelh reviewed (changes requested):
Generally LGTM, thanks!

andrei-edward-popa commented:
> Neat! FYI, I started a driver at https://github.com/nzmichaelh/zephyr/commit/1c5517654e23dd8faa07da0367badbf899ebbebf but never sent it upstream. Feel free to take anything you want from it.

@nzmichaelh Thank you for sharing this, I will take a look into it on sunday to see if it helps. <3

andrei-edward-popa commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+	gpio_pin_set_dt(&config->scl, 1);
+	gpio_pin_set_dt(&config->sda, 1);
+
+	while (!gpio_pin_get_dt(&config->scl)) {
```
This is another problem that I encountered with the I2C chip. It seems that SDA or SCL lines or both can be low when the device is powered up and it happens randomly. I searched for this problem on the internet and I found out that this I2C chip is maybe 90% a clone of a chip from STM32F1/4xx something. They have in their errata sheet an interesting part that SDA or SCL lines can be low instead of high at power up because of some analog filter inside the chip and they described in the errata sheet the steps to be implemented to not have this issue. The function wch_i2c_reset implements the exact same steps that they described and it seems to work for this I2C chip also.

nzmichaelh replied:
Understood. Could you add a comment along those lines and add a timeout.

The reason behind the timeout here and elsewhere is that any I2C peripheral is a different failure domain to the  CPU. So, for example, if there's a sensor on a plug, and the plug falls out, then the CPU should still start up and get a runtime error instead of failing to boot.

It's different for things in the same failure domain - for example, there's no real need to have a timeout when waiting for the internal ADC to finish a sample, as if the ADC is broken then the CPU is likely broken.

andrei-edward-popa replied:
Okay, thank you so much for helping me to understand this. Firstly, I need to test with external pull-up resistors to see if this reset function is required or not and if it works 100% of the time with the pull-ups I will remove it, else I will add the timeout.

andrei-edward-popa replied:
Outdated: I removed the reset function because using pull-ups works 100% of the time

VynDragon commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+
+	regs->CTLR1 &= ~I2C_CTLR1_PE;
+
+	gpio_pin_configure_dt(&config->scl, GPIO_OUTPUT | config->scl.dt_flags);
```
Yea that's is what i mean is a problem, if there is a issue with pinctrl, mitigation needed or not, it needs to be pinctrl, not here, and especially not need scl and sda given to the i2c driver.

andrei-edward-popa replied:
Read my comment about the wch_i2c_reset function. The i2c pins are configured through pinctrl and it is not a problem with it (I think). This function is here because the analog filter can cause SDA or SCL to be low at power up randomly and they can be recovered using this method (see STM32 errata sheet for STM32F1/4xx what is this about).

VynDragon replied:
The point is that you should get i2c pins from bus if you really need them, not as a peripheral argument.

andrei-edward-popa replied:
Is there an API to do that? To get gpio_dt_spec using pinctrl?

VynDragon replied:
I believe that would be through PINCTRL_DT_DEV_CONFIG_GET(node_id)

andrei-edward-popa replied:
Outdated: I removed the reset function because using pull-ups works 100% of the time

andrei-edward-popa commented:
@VynDragon Do you have external pull up resistors on sda and scl? I read the AFIO documentation and I think they are required because pull up or pull down can be activated only on input mode. I didn't have external pull up resistors when I tested the driver because I thought they should be enabled with bias-pull-up, but on output mode it has no effect as I understand. My question is: are external pull up resistors required as per your understanding?

VynDragon commented:
> @VynDragon Do you have external pull up resistors on sda and scl? I read the AFIO documentation and I think they are required because pull up or pull down can be activated only on input mode. I didn't have external pull up resistors when I tested the driver because I thought they should be enabled with bias-pull-up, but on output mode it has no effect as I understand. My question is: are external pull up resistors required as per your understanding?

Definitely no pull-ups on the board.
Pins are not in input or output modes for i2c, but in alternate mode, and it says this:
``` To use the alternate function of the input direction, the port must be configured to alternate input mode,
the pull-up and pull-down settings can be set according to actual needs```

andrei-edward-popa commented:
> > @VynDragon Do you have external pull up resistors on sda and scl? I read the AFIO documentation and I think they are required because pull up or pull down can be activated only on input mode. I didn't have external pull up resistors when I tested the driver because I thought they should be enabled with bias-pull-up, but on output mode it has no effect as I understand. My question is: are external pull up resistors required as per your understanding?
> 
> Definitely no pull-ups on the board.
> I think it's just neverending stop condition, it works with SSD1306 without resistors on ch32fun on ch32v003 which has the same gpio schematics.

Can you add pull-ups and test on 003 and 203 also? I'm on a trip and I can't do it until monday :( Maybe on 003 it just happens to work without the pull-ups

VynDragon commented:
> Can you add pull-ups and test on 003 and 203 also? I'm on a trip and I can't do it until monday :( Maybe on 003 it just happens to work without the pull-ups
No difference on 203 (it put a 100K, it's just not in open drain mode). 003 still cant any of those samples due to lack of space and memory.

andrei-edward-popa commented:
> > Can you add pull-ups and test on 003 and 203 also? I'm on a trip and I can't do it until monday :( Maybe on 003 it just happens to work without the pull-ups
> 
> No difference on 203 (i put a 100K, it's just not getting back into open drain mode). 003 still cant any of those samples due to lack of space and memory.
> 

I think it is too high, do you have some between 2k and 10k?

VynDragon commented:
> > > Can you add pull-ups and test on 003 and 203 also? I'm on a trip and I can't do it until monday :( Maybe on 003 it just happens to work without the pull-ups
> > 
> > 
> > No difference on 203 (i put a 100K, it's just not getting back into open drain mode). 003 still cant any of those samples due to lack of space and memory.
> 
> I think it is too high, do you have some between 2k and 10k?

It should not be too high if the pin was going back to open drain like it should. Exact same result with 2K2.

andrei-edward-popa commented:
> > > Can you add pull-ups and test on 003 and 203 also? I'm on a trip and I can't do it until monday :( Maybe on 003 it just happens to work without the pull-ups
> > 
> > No difference on 203 (i put a 100K, it's just not getting back into open drain mode). 003 still cant any of those samples due to lack of space and memory.
> > 
> 
> I think it is too high, do you have some between 2k and 10k?

Okay, thank you so much for testing. I will do more testing when I'm back and let you know other findings.

andrei-edward-popa commented on dts/riscv/wch/ch32v0/ch32v003.dtsi:
```diff
@@ -102,6 +102,16 @@
+			reg = <0x40005400 0x400>;
+			clocks = <&rcc CH32V00X_CLOCK_I2C1>;
+			interrupt-parent = <&pfic>;
+			interrupts = <30>,<31>;
```
Done

andrei-edward-popa commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+{
+	int ret = 0;
+
+	msg[0].flags |= I2C_MSG_RESTART;
```
I passed to the begin transfer function if it is the first message or not.

andrei-edward-popa commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+	for (uint8_t i = 0; i < num_msgs; i++) {
+		ret = wch_i2c_begin_transfer(dev, &msg[i], addr);
+		if (ret < 0) {
+			break;
```
Done

andrei-edward-popa commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,498 @@
+	while (!gpio_pin_get_dt(&config->scl)) {
+	}
+
+	while (!gpio_pin_get_dt(&config->sda)) {
```
Outdated: I removed the reset function because using pull-ups works 100% of the time

andrei-edward-popa commented:
Hi @VynDragon, can you check again on ch32v203, please?

andrei-edward-popa commented:
Hi @VynDragon @miggazElquez, please see https://github.com/zephyrproject-rtos/zephyr/issues/88449. It is not related to i2c driver directly, but this driver uses interrupts and after some iterations of writing/reading through i2c is hangs because of the mentioned issue.

VynDragon commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,433 @@
+		trise = freq_range + 1;
+		break;
+	case I2C_SPEED_FAST:
+		trise = (freqrange * 3 / 10) + 1;
```
Typo freq_range

andrei-edward-popa replied:
Sorry for that, it is resolved now.

VynDragon reviewed (changes requested):
Current status: Does not output data anymore
![image](https://github.com/user-attachments/assets/1c1105b7-e935-4cb3-bb1d-982345ea936d)

andrei-edward-popa commented:
> Current status: Does not output data anymore
> ![image](https://github.com/user-attachments/assets/1c1105b7-e935-4cb3-bb1d-982345ea936d)
> 

Okay, let's have the interrupt issue solbed first if it reproduces on ch32v203 also because I think it is related to what are you see and if it doesn't work after that I will test with the board, it is expected to come in 7 days.

VynDragon commented:
Do not worry about going fast, PRs usually take a very long time to be merged

andrei-edward-popa commented:
@VynDragon Hi, I get the board with CH32V203 and indeed the I2C was not working on it. I found the problem and I solved it. I tested on CH32V003 and CH32V203 with 3 I2C devices and on CH32V203 with an additional one, the ssd1306 display and it works. Can you check it?

VynDragon commented:
Yes it works :+1:

VynDragon commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,441 @@
+		}
+	}
+
+	regs->STAR1;
```
Still not sure what is the idea there, but if i remove them the driver stops working, while this technically a no-op? I would get it if it was a read op... Would you mind explaining it to me?

VynDragon replied:
Ah ok, i am confused why this is doing a read with no attribution, i suppose it must be a ch32fun subtelty.

VynDragon reviewed (approved):
It works and I dont see any issue.

VynDragon commented:
I wanted to try with a different display, pulled SSD1363 out of the stack and it's having load address misaligned issue, currently debugging this to see if which driver / bit is having the issue (could be that a buffer needs to be aligned).

andrei-edward-popa commented:
> It fails in multithreading start, not too sure what it is unhappy about, but multithreading cannot be disabled due to the semaphore in this code. I will try a different display (a less 'heavy' one) and see if the issue happens still.

Can you increase the stack size of the main thread and isr? For CH32V203 it can be much bigger.

VynDragon commented:
~~I tried a different display (one that wasnt hitting as close to ram limits, driver arch is still almost the same, part of my serie of oled drivers) and it's fine. ~~
Found the issue, entirely my fault, broke the driver on my last push with a bad copy-paste :facepalm: 
![image](https://github.com/user-attachments/assets/e357ab3e-5747-43b2-9fa0-74ec51ca9191)
I2c driver is all good :+1:

andrei-edward-popa commented:
> ~~I tried a different display (one that wasnt hitting as close to ram limits, driver arch is still almost the same, part of my serie of oled drivers) and it's fine.~~
> Found the issue, entirely my fault, broke the driver on my last push with a bad copy-paste :facepalm: 
> ![image](https://github.com/user-attachments/assets/e357ab3e-5747-43b2-9fa0-74ec51ca9191)
> I2c driver is all good :+1: (this specific display browns out ch32v203's power supply instead now :) )
> 

I see, I'm glad that the I2C driver is working fine now on all SoCs 😊

VynDragon commented:
Nevermind, I am now hitting issue with BME280 sample that i forgot to test.
It gets to 'cfg->bus_io->write(&cfg->bus, reg, val);', then crashes right there. with illegal instruction. I checked all values it should all be correct. 

![image](https://github.com/user-attachments/assets/9dc0f45f-53d4-4b8d-a1d9-04eb8cf9321a)

I will try and see if it happens on ch32v003, it should not be crashing there.

VynDragon commented:
So i was testing BME280 and turns out it's not good for testing anymore, they changed it to use async API and it has no checks for whether i2c supports the async mode...

andrei-edward-popa commented:
> So i was testing with BME280 for reads and turns out it's not good for testing anymore, they changed it to use async API and it has no checks for whether i2c supports the async mode... Which resulted in another unexplainable crash...

Oh, I see :( For reads I tested with ds3231 (RTC) and mpu6050 (accelerometer + gyro + temp) and it works on both SoCs that I have. I tested 1 byte read, 2 bytes read and >=3 bytes read to see if the stop condition is generated ok for "corner cases" like 1 byte and 2 bytes (in manual tney said that some things must be done if we want to receive 1 or 2 bytes) and it works on all of this cases.

VynDragon commented:
It's all good, after spending some time on it and asking around, it turned out to be a stack issue, and the possible async issue isnt present either (apparently it should?).

The fix was to remove  the stack size kconfigs in Kconfig.defconfig of ch32v203.

andrei-edward-popa commented on drivers/i2c/i2c_wch.c:
```diff
@@ -0,0 +1,441 @@
+
+static DEVICE_API(i2c, i2c_wch_api) = {
+	.configure = i2c_wch_configure,
+	.transfer = i2c_wch_transfer,
```
Done

andrei-edward-popa commented:
> It's all good, after spending some time on it and asking around, it turned out to be a ch32v203 stack issue (like nzmichael warned about actually...), and the possible async issue isnt present either (apparently it should?).
> 
> The fix was to remove the stack size kconfigs in Kconfig.defconfig of ch32v203.
> 
> And yes BME280 reads work perfectly.

That's great to know. Thank you for all of your testing, I'm very glad that the driver works now. :)

VynDragon reviewed (approved):
Do not push rebase at the same time as changes, it makes it impossible to tell easily what changed.

andrei-edward-popa commented:
> Do not push rebase at the same time as changes, it makes it impossible to tell easily what changed.

Okay, good to know, thank you for all! 😊
